### PR TITLE
Add request body streaming. Fixes #622

### LINKED ISSUE #622: Streaming request body
Hello! First of all, thanks for the awesome work.

I'm building a simple file server that receives files from clients. The files are presented in the raw request body (i.e. no multipart/form-data), and file sizes can be several hundred megabytes large. fasthttp seems to have a small default request body limit (4MB), and when I try to increase that limit via MaxRequestBodySize I get out of memory error when uploading large files. Is there a way to stream request body directly so it uses only constant memory?

### PR DESCRIPTION
**Pull request by kirillDanshin**

### CODE REVIEW

File: http.go [hunk 1]
```diff
@@ -1017,6 +1021,73 @@ func (req *Request) readLimitBody(r *bufio.Reader, maxBodySize int, getOnly bool
...
+		req.Header.SetContentLength(0)
+		return nil
+	}
+
+	bodyBuf := req.bodyBuffer()
+	bodyBuf.Reset()
+	bodyBuf.B, err = readBodyWithStreaming(r, contentLength, maxBodySize, bodyBuf.B)
+	if err != nil {
+		if err == ErrBodyTooLarge {
+			req.Header.SetContentLength(contentLength)
```
Comment (erikdubbelboer) [inline http.go:1068]: Shouldn't this be `maxBodySize`?

File: http.go [hunk 2]
```diff
@@ -1017,6 +1021,73 @@ func (req *Request) readLimitBody(r *bufio.Reader, maxBodySize int, getOnly bool
...
+	}
+
+	var err error
+	contentLength := req.Header.realContentLength()
+	if contentLength > 0 {
+		if preParseMultipartForm {
+			// Pre-read multipart form data of known length.
+			// This way we limit memory usage for large file uploads, since their contents
+			// is streamed into temporary files if file size exceeds defaultMaxInMemoryFileSize.
+			req.multipartFormBoundary = string(req.Header.MultipartFormBoundary())
```
Comment (ReneWerner87) [inline http.go:1043]: should lead to allowcations, please check again

File: http.go [hunk 3]
```diff
@@ -1081,6 +1152,69 @@ func (req *Request) ContinueReadBody(r *bufio.Reader, maxBodySize int, preParseM
...
+// then ErrBodyTooLarge is returned.
+func (req *Request) ContinueReadBodyStream(r *bufio.Reader, maxBodySize int, preParseMultipartForm ...bool) error {
+	var err error
+	contentLength := req.Header.realContentLength()
+	if contentLength > 0 {
+		if len(preParseMultipartForm) == 0 || preParseMultipartForm[0] {
+			// Pre-read multipart form data of known length.
+			// This way we limit memory usage for large file uploads, since their contents
+			// is streamed into temporary files if file size exceeds defaultMaxInMemoryFileSize.
+			req.multipartFormBoundary = string(req.Header.MultipartFormBoundary())
```
Comment (ReneWerner87) [inline http.go:1170]: should lead to allowcations, please check again

File: requestStream.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	"io"
+	"sync"
+
+	"github.com/valyala/bytebufferpool"
+)
+
+type requestStream struct {
+	prefetchedBytes *bytes.Reader
+	reader          *bufio.Reader
+	n               int
```
Comment (ReneWerner87) [inline requestStream.go:16]: please add a better name for this

File: http.go [hunk 5]
```diff
@@ -1017,6 +1021,73 @@ func (req *Request) readLimitBody(r *bufio.Reader, maxBodySize int, getOnly bool
...
+	bodyBuf.B, err = readBodyWithStreaming(r, contentLength, maxBodySize, bodyBuf.B)
+	if err != nil {
+		if err == ErrBodyTooLarge {
+			req.Header.SetContentLength(contentLength)
+			req.body = bodyBuf
+			req.bodyRaw = bodyBuf.B[:maxBodySize]
+			req.bodyStream = acquireRequestStream(bodyBuf, r, contentLength)
+			return nil
+		}
+		if err == errChunkedStream {
```
Comment (ReneWerner87) [inline http.go:1074]: i think "switch-case" is better for this

File: http.go (hunk 1 above)
Comment (kirillDanshin) [reply http.go:1068]: With streams, now you can set body limit to 1 and still read ContentLenght worth of data if you'd like. How you supposed to know content length if we set the header to 1? 🤔

**Comment by Fenny** [comment]
```go
func main() {
    app := &fasthttp.Server{
        Handler:            handler,
        StreamRequestBody:  true,
        MaxRequestBodySize: 1,
    }
    log.Fatal(app.ListenAndServe(":3000"))
}

func handler(ctx *fasthttp.RequestCtx) {
    var (
        stream = ctx.RequestBodyStream()
        read   = 0
        buf    = make([]byte, 1024*1024) // 1MB
    )

    for {
        n, err := stream.Read(buf)
        read += n

        if err == io.EOF {
            break
        }

        if err != nil {
            log.Fatal(err.Error())
        }
    }

    fmt.Println("Read: ", read, "bytes")
}
```
6MB Ram, reading a 1.8GB body size 👍

File: requestStream.go (hunk 4 above)
Comment (Thor-x86) [reply requestStream.go:16]: like `streamRequest`?

**Review by Thor-x86 — APPROVED** [review]
`StreamRequestBody` is really helpful for storage server with authorization like JWT and OAUTH without splitting and merging files on client side. Thus, I approve this change.

**Comment by frankdvd** [comment]
Any idea on when this will be merged? This feature is really helpful, we really need it. Thank you for the works!!!!!

**Comment by juliangp** [comment]
Is there any reason why this is still in draft, we need this feature and are ready to help if needed?

File: http.go [hunk 6]
```diff
@@ -1081,6 +1132,73 @@ func (req *Request) ContinueReadBody(r *bufio.Reader, maxBodySize int, preParseM
...
+// then ErrBodyTooLarge is returned.
+func (req *Request) ContinueReadBodyStream(r *bufio.Reader, maxBodySize int, preParseMultipartForm ...bool) error {
+	var err error
+	contentLength := req.Header.realContentLength()
+	if contentLength > 0 {
+		if len(preParseMultipartForm) == 0 || preParseMultipartForm[0] {
+			// Pre-read multipart form data of known length.
+			// This way we limit memory usage for large file uploads, since their contents
+			// is streamed into temporary files if file size exceeds defaultMaxInMemoryFileSize.
+			req.multipartFormBoundary = string(req.Header.MultipartFormBoundary())
```
Comment (erikdubbelboer) [inline http.go:1150]:
```suggestion
			req.multipartFormBoundary = b2s(req.Header.MultipartFormBoundary())
```

File: http.go [hunk 7]
```diff
@@ -1017,6 +1021,53 @@ func (req *Request) readLimitBody(r *bufio.Reader, maxBodySize int, getOnly bool
...
+	}
+
+	var err error
+	contentLength := req.Header.realContentLength()
+	if contentLength > 0 {
+		if preParseMultipartForm {
+			// Pre-read multipart form data of known length.
+			// This way we limit memory usage for large file uploads, since their contents
+			// is streamed into temporary files if file size exceeds defaultMaxInMemoryFileSize.
+			req.multipartFormBoundary = string(req.Header.MultipartFormBoundary())
```
Comment (erikdubbelboer) [inline http.go:1043]:
```suggestion
			req.multipartFormBoundary = b2s(req.Header.MultipartFormBoundary())
```

File: requestStream_test.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	println("written the request")
+	println(formattedRequest)
+	br := bufio.NewReader(c)
+	var respH ResponseHeader
+	if err = respH.Read(br); err != nil {
+		t.Errorf("unexpected error: %s", err)
+	}
+
+	println(respH.String())
+	// time.Sleep(5 * time.Second)
```
Comment (erikdubbelboer) [inline requestStream_test.go:89]: Lets not commit commented out code. Also can you use `t.Log` instead of `println`.

File: requestStream.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
+package fasthttp
```
Comment (erikdubbelboer) [inline requestStream.go:1]: Can we rename this file to `streaming.go`. We always lowercase all files.

File: requestStream_test.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
+package fasthttp
```
Comment (erikdubbelboer) [inline requestStream_test.go:1]: `streaming_test.go`

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

**Review by erikdubbelboer — APPROVED** [review]

**Review by erikdubbelboer — APPROVED** [review]

**Pull request merged**